File: 0133-clone-graph/0133-clone-graph.java
/*
// Definition for a Node.
class Node {
  public int val;
  public List<Node> neighbors;
  public Node() {
    val = 0;
    neighbors = new ArrayList<Node>();
  }
  public Node(int _val) {
    val = _val;
    neighbors = new ArrayList<Node>();
  }
  public Node(int _val, ArrayList<Node> _neighbors) {
    val = _val;
    neighbors = _neighbors;
  }
}
*/

class Solution {
  public void dfs(Node node, Map<Integer, Node> newGraph, Set<Integer> visited) {
    visited.add(node.val);
    if (!newGraph.containsKey(node.val)) newGraph.put(node.val, new Node(node.val));
    for (Node nb : node.neighbors) {
      if (!newGraph.containsKey(nb.val)) newGraph.put(nb.val, new Node(nb.val));
      if (!newGraph.get(node.val).neighbors.contains(newGraph.get(nb.val))) {
        newGraph.get(node.val).neighbors.add(newGraph.get(nb.val));
      }
      if (!visited.contains(nb.val)) dfs(nb, newGraph, visited);
    }
  }
  
  public Node cloneGraph(Node node) {
    Map<Integer, Node> newGraph = new HashMap();
    if (node == null) return null;
    Set<Integer> visited = new HashSet<Integer>();
    dfs(node, newGraph, visited);
    return newGraph.get(1);
  }
}

File: 0133-clone-graph/0133-clone-graph.go
/**
 * Definition for a Node.
 * type Node struct {
 *   Val int
 *   Neighbors []*Node
 * }
 */

func cloneGraph(node *Node) *Node {
  newGraph, visited := make(map[int]*Node), make(map[int]bool)
  if node == nil {
    return nil
  }
  dfs(node, &newGraph, &visited)
  return newGraph[1]
}

func dfs(node *Node, newGraph *map[int]*Node, visited *map[int]bool)  {
  (*visited)[(*node).Val] = true
  if _, ok := (*newGraph)[(*node).Val]; !ok {
    (*newGraph)[(*node).Val] = &Node{
      Val: (*node).Val,
    }
  }
  for _, nb := range (*node).Neighbors {
    if _, ok := (*newGraph)[(*nb).Val]; !ok {
      (*newGraph)[(*nb).Val] = &Node{
        Val: (*nb).Val,
      }
    }
    if !slices.Contains((*(*newGraph)[(*node).Val]).Neighbors, (*newGraph)[(*nb).Val]) {
      (*(*newGraph)[(*node).Val]).Neighbors = append((*(*newGraph)[(*node).Val]).Neighbors, (*newGraph)[(*nb).Val])
    }
    if _, ok := (*visited)[(*nb).Val]; !ok {
      dfs(nb, newGraph, visited)
    }
  }
}